lib/pos_fee_estimator.go

```diff
@@ -155,5 +180,149 @@ func (posFeeEstimator *PoSFeeEstimator) mempoolFeeEstimate(
+	}
+	const MaxDERSigLen = 74
+	txnBytesLen := uint64(len(txnBytesNoSignature)) + MaxDERSigLen
+
+	txnFee := txnBytesLen * feeRateNanosPerKB / 1000
+
+	if (txnBytesLen*feeRateNanosPerKB)%1000 != 0 {
+		txnFee++
+	}
+	if txnFee < txn.TxnFeeNanos {
+		txn.TxnFeeNanos = txnFee
+		return computeFeeRecursive(txn, feeRateNanosPerKB)
+	}
+	return txnFee, nil
+}
+
+func (posFeeEstimator *PoSFeeEstimator) estimateTxnFeeGivenTransactionRegister(
+	txn *MsgDeSoTxn,
+	txnRegister *TransactionRegister,
+	congestionFactorBasisPoints uint64,
+	priorityPercentileBasisPoints uint64,
+	maxBlockSize uint64,
+) (uint64, error) {
+	feeRateNanosPerKB, err := posFeeEstimator.estimateFeeRateNanosPerKBGivenTransactionRegister(txnRegister, congestionFactorBasisPoints, priorityPercentileBasisPoints, maxBlockSize)
+	if err != nil {
+		return 0, errors.Wrap(err, "estimateTxnFeeGivenTransactionRegister: Problem computing fee rate")
+	}
+	txnFee, err := computeFeeGivenTxnAndFeeRate(txn, feeRateNanosPerKB)
+	if err != nil {
+		return 0, errors.Wrap(err, "estimateTxnFeeGivenTransactionRegister: Problem computing txn fee")
+	}
+	return txnFee, nil
+}
+
+func (posFeeEstimator *PoSFeeEstimator) estimateFeeRateNanosPerKBGivenTransactionRegister(
+	txnRegister *TransactionRegister,
+	congestionFactorBasisPoints uint64,
+	priorityPercentileBasisPoints uint64,
+	maxBlockSize uint64,
+) (uint64, error) {
+	it := txnRegister.GetFeeTimeIterator()
+	maxSizeOfNumBlocks := maxBlockSize * posFeeEstimator.numBlocks
+	totalTxnsSize := uint64(0)
+	var txns []*MempoolTx
+	for it.Next() {
+		tx, ok := it.Value()
+		if !ok {
+			break
+		}
+		txnBytes, err := tx.Tx.ToBytes(false)
+		if err != nil {
+			return 0, errors.Wrap(err, "estimateFeeRateNanosPerKBGivenTransactionRegister: Problem serializing txn")
+		}
+		totalTxnsSize += uint64(len(txnBytes))
+		txns = append(txns, tx)
+		// TODO: I think we want to include the txn that puts us over the limit, but
+		// we can just move this check up a few lines if that's wrong.
+		if totalTxnsSize > maxSizeOfNumBlocks {
+			break
+		}
+	}
+	// TODO: global min fee? or is 0 okay?
+	if len(txns) == 0 {
+		return 0, nil
+	}
+
+	// Compute the congestion threshold. If our congestion factor is 100% (or 10,000 bps),
+	// then congestion threshold is simply max block size * numBlocks
+	// TODO: I don't know if I like this name really.
+	congestionThreshold := (congestionFactorBasisPoints * posFeeEstimator.numBlocks * maxBlockSize) / (100 * 100)
+	bucketMinFee, bucketMaxFee, err := getPriorityFeeBucketFromTxns(txnRegister, txns, priorityPercentileBasisPoints)
+	if err != nil {
+		return 0, errors.Wrap(err, "estimateFeeRateNanosPerKBGivenTransactionRegister: Problem computing priority fee bucket")
+	}
+	// If the bucketMinFee is 0, we simply return 0.
+	if bucketMinFee == 0 {
+		// Ugh I don't like this.
+		globalMinFeeRate, _ := txnRegister.minimumNetworkFeeNanosPerKB.Uint64()
+		return globalMinFeeRate, nil
```

Comment:
I agree that this is the least bad approach. You can make it slightly more robust by clamping explicitly to the network's min fee rather than doing a zero check. Seems kosher
```suggestion
	globalMinFeeRate, _ := txnRegister.minimumNetworkFeeNanosPerKB.Uint64()

	// If the bucketMinFee is 0, we simply return 0.
	if bucketMinFee < globalMinFeeRate {
		return globalMinFeeRate, nil
```